pkg/cloudprovider/overlay/cloudprovider.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/samber/lo"
+	lop "github.com/samber/lo/parallel"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	v1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/apis/v1alpha1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/operator/options"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+)
+
+type decorator struct {
+	cloudprovider.CloudProvider
+	kubeClient client.Client
+}
+
+// Decorate returns a new `CloudProvider` instance that will delegate the GetInstanceTypes
+// calls to the argument, `cloudProvider`, and provide instance types with NodeOverlays applied to them. The
+func Decorate(cloudProvider cloudprovider.CloudProvider, kubeClient client.Client) cloudprovider.CloudProvider {
+	return &decorator{CloudProvider: cloudProvider, kubeClient: kubeClient}
+}
+
+func (d *decorator) GetInstanceTypes(ctx context.Context, nodePool *v1.NodePool) ([]*cloudprovider.InstanceType, error) {
+	its, err := pullInstanceTypes(ctx, d.CloudProvider, nodePool)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+	if options.FromContext(ctx).FeatureGates.NodeOverlay {
+		// The additional requirements will be added to the instance type during scheduling simulation
+		// Since getting instance types is done on a NodePool level, these requirnements were always assumed
+		// to be allowed with these instance types.
+		addNodePoolRequirements(nodePool, its)
+		its, err = OverlayInstanceTypes(ctx, d.kubeClient, its)
+		if err != nil {
+			return []*cloudprovider.InstanceType{}, err
+		}
+		removeNodePoolRequirements(nodePool, its)
+	}
+	return its, nil
+}
+
+func OverlayInstanceTypes(ctx context.Context, kubeClient client.Client, instanceTypes []*cloudprovider.InstanceType) ([]*cloudprovider.InstanceType, error) {
+	overlays := &v1alpha1.NodeOverlayList{}
+	err := kubeClient.List(ctx, overlays)
+	if err != nil {
+		return []*cloudprovider.InstanceType{}, err
+	}
+
+	// Only apply overlays that have passed runtime validation and know there have not been
+	// any updates to the spec of the overlay field
+	validOverlays := lo.Filter(overlays.Items, func(o v1alpha1.NodeOverlay, _ int) bool {
+		condition := o.StatusConditions().Get(v1alpha1.ConditionTypeValidationSucceeded)
+		return condition.IsTrue() && condition.ObservedGeneration == o.Generation
+	})
+
+	// We order the overlays from largest to smallest, as we have already done validation
+	// we can be sure that we will not be blocked due to invalid or conflicting overlays
+	sort.Slice(validOverlays, func(x int, y int) bool {
+		return lo.FromPtr(validOverlays[x].Spec.Weight) > lo.FromPtr(validOverlays[y].Spec.Weight)
+	})
+	work := []func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType){
+		overlayCapacityOnInstanceTypes,
+		overlayPriceOnInstanceTypes,
+	}
+	lop.ForEach(work, func(f func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType), i int) {
+		f(validOverlays, instanceTypes)
+	})
```

Comment:
nit: we don't really use this `ForEach` helper elsewhere, and I'm not sure it provides much value over a native for each loop. What I've found in practice is it just makes debugging more difficult due to the additional closure
```suggestion
	for _, f := range []func(overlays []v1alpha1.NodeOverlay, its []*cloudprovider.InstanceType){
		overlayCapacityOnInstanceTypes,
		overlayPriceOnInstanceTypes,
	} {
	    f(validOverlays, instanceTypes)
	}
```

Reply:
I would consider collapsing these two functions though since it requires we double our instance compatibility checks.

Reply:
This function is using `lop "github.com/samber/lo/parallel"` runs both function in parallel. You could do that in a for loop, but it made sense to parallelize  the call.  I do think it's fine to just combine these functions.